uc.c

```diff
@@ -505,49 +519,158 @@ uc_err uc_close(uc_engine *uc)
+            return err;
+        }
+    }
+
+    return UC_ERR_OK;
+}
+
+UNICORN_EXPORT
+uc_err uc_reg_write_batch(uc_engine *uc, int *regs, void *const *vals,
+                          int count)
+{
     UC_INIT(uc);
+    reg_write_t reg_write = uc->reg_write;
+    void *env = uc->cpu->env_ptr;
+    int mode = uc->mode;
+    int setpc = 0;
+    int i;
 
-    if (uc->reg_read) {
-        ret = uc->reg_read(uc, (unsigned int *)ids, vals, count);
-    } else {
-        return UC_ERR_HANDLE;
+    for (i = 0; i < count; i++) {
+        unsigned int regid = regs[i];
+        const void *value = vals[i];
+        size_t size = (size_t)-1;
+        uc_err err = reg_write(env, mode, regid, value, &size, &setpc);
+        if (err) {
+            return err;
+        }
+    }
+    if (setpc) {
+        // force to quit execution and flush TB
+        uc->quit_request = true;
+        break_translation_loop(uc);
     }
 
-    return ret;
+    return UC_ERR_OK;
 }
 
 UNICORN_EXPORT
-uc_err uc_reg_write_batch(uc_engine *uc, int *ids, void *const *vals, int count)
+uc_err uc_reg_read_batch2(uc_engine *uc, int *regs, void *const *vals,
+                          size_t *sizes, int count)
 {
-    int ret = UC_ERR_OK;
+    UC_INIT(uc);
+    reg_read_t reg_read = uc->reg_read;
+    void *env = uc->cpu->env_ptr;
+    int mode = uc->mode;
+    int i;
+
+    for (i = 0; i < count; i++) {
+        unsigned int regid = regs[i];
+        void *value = vals[i];
+        uc_err err = reg_read(env, mode, regid, value, sizes + i);
+        if (err) {
+            return err;
+        }
+    }
 
+    return UC_ERR_OK;
+}
+
+UNICORN_EXPORT
+uc_err uc_reg_write_batch2(uc_engine *uc, int *regs, const void *const *vals,
+                           size_t *sizes, int count)
+{
     UC_INIT(uc);
+    reg_write_t reg_write = uc->reg_write;
+    void *env = uc->cpu->env_ptr;
+    int mode = uc->mode;
+    int setpc = 0;
+    int i;
 
-    if (uc->reg_write) {
-        ret = uc->reg_write(uc, (unsigned int *)ids, vals, count);
-    } else {
-        return UC_ERR_HANDLE;
+    for (i = 0; i < count; i++) {
+        unsigned int regid = regs[i];
+        const void *value = vals[i];
+        uc_err err = reg_write(env, mode, regid, value, sizes + i, &setpc);
+        if (err) {
+            return err;
+        }
+    }
+    if (setpc) {
+        // force to quit execution and flush TB
+        uc->quit_request = true;
+        break_translation_loop(uc);
     }
 
-    return ret;
+    return UC_ERR_OK;
 }
 
 UNICORN_EXPORT
 uc_err uc_reg_read(uc_engine *uc, int regid, void *value)
 {
     UC_INIT(uc);
-    return uc_reg_read_batch(uc, &regid, &value, 1);
+    size_t size = (size_t)-1;
+    return uc->reg_read(uc->cpu->env_ptr, uc->mode, regid, value, &size);
 }
 
 UNICORN_EXPORT
 uc_err uc_reg_write(uc_engine *uc, int regid, const void *value)
 {
     UC_INIT(uc);
-    return uc_reg_write_batch(uc, &regid, (void *const *)&value, 1);
+    int setpc = 0;
+    size_t size = (size_t)-1;
```

Comment:
Personally I perfer using `0` as the default value because `(size_t)-1` implies the maximum value of the unsigned type, which is error-prone when writing `reg_read` code. For example, some contributors may simply write `memcpy(val, ptr, size)` someday without noticing that `size` can be `(size_t)-1`. However, this requires some extra code when checking the size but I think the tradeoff worths it.

Reply:
The precise semantics of `size` are that it specifies the size of the buffer on input, and the size that was actually written on output. It is completely reasonable that callers may allocate more memory than required, and they can check how much memory was actually used via the return value in `*size`. `memcpy(val, ptr, size)` is categorically wrong, because the size is just a maximum. I would call it `maxsize`, but because it gets set to the true size upon return, that would be slightly misleading.

`(size_t)-1` is for compatibility with the old API, which assumed that the buffer was unbounded. `(size_t)0` will require special handling and is semantically more convoluted. The most straightforward definition is that `(size_t)0` means a buffer of size zero; although this seems useless on the surface, it can be useful for e.g. poking registers that have side-effects (none in current architectures, but many ARM CP registers behave this way).

Reply:
Sound reasonable!